R/data-storage.R

```diff
@@ -96,37 +92,90 @@ DataStorage <- R6::R6Class( # nolint object_name_linter
         ))
       })
     },
-    generate_session_id = function() {
-      paste(
-        c(
-          sample(c(letters, LETTERS, 0:9), 10),
-          format(Sys.time(), "%Y%d%m%H%M%S")
-        ),
-        collapse = ""
-      )
-    },
+
     close_connection = function() {
       rlang::abort("Method not implemented.")
     },
-    insert_checks = function(values, insert_time, bucket) {
-      checkmate::assert_string(bucket)
-      checkmate::assert_list(values)
-      checkmate::assert_flag(insert_time)
 
-      if (!insert_time) {
-        return(values)
-      }
+    write = function(values, bucket) {
+      rlang::abort("Method not implemented.")
+    },
 
-      if ("time" %in% names(values)) {
-        rlang::abort(paste0(
-          "You must not pass 'time' value into database.",
-          " It is set automatically."
-        ))
-      }
+    read_data = function(date_from, date_to, bucket) {
+      rlang::abort("Method not implemented.")
+    },
 
-      values$time <- as.character(Sys.time())
+    insert_checks = function(app_name, type, session, details, time) {
+      checkmate::assert_string(app_name)
+      checkmate::assert_string(type, null.ok = TRUE)
+      checkmate::assert_string(session, null.ok = TRUE)
+      checkmate::assert_class(time, "POSIXct", null.ok = TRUE)
+      checkmate::assert(
+        .combine = "or",
+        checkmate::check_scalar(details),
+        checkmate::check_list(details, null.ok = TRUE)
+      )
+
+      values <- list(
+        app_name = app_name,
+        type = type,
+        session = session,
+        details = details
+      )
+
+      if (checkmate::test_list(details)) {
+        values$details <- jsonlite::toJSON(details)
+      }
 
+      values$time <- dplyr::coalesce(time, Sys.time())
+      values$time <- as.character(values$time)
       values
+    },
+
+    # Unnest JSON column from data.frame
+
+    unnest_json = function(x, column_name) {
```

Comment:
Convert column in table to multiple columns.

`fromJSON()` does not accept vector input, so we need to map it. There might be performance concerns here.

I'm VERY open to alternate solutions!!